### PR TITLE
FED-2156 Unsafe required props lint

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
While required late props are validated to be present when a component is rendered, there are cases where they won't always be present. For example, "partial" props objects that represent a subset of a component's props.

If required late props are accessed without being set, they'll throw at runtime, and there's currently no static analysis that prevents consumers from doing so, or notifies them that they're doing something unsafe.

So, we want to help prevent consumers from unsafely accessing required late props.

## Changes
- Add new `UnsafeRequiredPropAccessDiagnostic` that warns whenever required props are accessed in an unsafe manner
    - Lints when accessing required props on objects that aren't props a component was rendered with
    - Exceptions where we don't lint
        - Accesses on callback args in utility methods like `.getPropKey((p) => p.requiredProp)`
        - Within `memo` areEqual functions, or top-level functions with `areEqual` in the name since consumers apparently split themout a fair bit
        - In files within the `test/` directory, which typically uses `getProps` a lot (and if unsafe accesses happen, they'll just fail tests instead of crashing an app)
        - Within lifecycle methods that receive props as an argument and typically use `typedPropsFactory`
            - We could track these down and ignore them only if they're safe, but that didn't seem worth the effort/complexity
- Add tests
   - Update `SharedAnalysisContext` to allow creation of files outside of `lib`
- Add playground example

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - Verify CI passes
        - Manually test
            - Test in playground example
            - Test in a consumer that uses required props (e.g., FluxUiProps) and verify there aren't any unexpected errors
            - Double-check with `// debug: over_react_metrics` that performance is acceptable (this lint has to process all property accesses to see if they're on props). From my testing it looked pretty efficient, all things considered, and was hovering around being our 5th slowest lint, which IMO isn't bad at all.
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by sydneyjodon-wk — APPROVED** [review]

**Review by sydneyjodon-wk — COMMENTED** [review]
This is super awesome, Greg! You were so thorough!!!!

The only thing that came up in QA was that on my computer, `// debug:over_react_metrics` in the playground has consistently shown `UnsafeRequiredPropAccessDiagnostic` to be one of the two slowest - at first it was always the slowest at around 44%, but now whenever I restart my analysis server, it's consistently section slowest:

```
  ** UnsafeRequiredPropAccessDiagnostic: 3.483ms (24.9%)
  ** UnsafeRequiredPropAccessDiagnostic: 3.270ms (25.2%),
  ** UnsafeRequiredPropAccessDiagnostic: 3.488ms (25.1%),
  ** UnsafeRequiredPropAccessDiagnostic: 3.344ms (25.3%),
```

I don't know if that's much different from what you were seeing since you said it was around fifth slowest?

QA steps:
- [x] Test in playground example
- [x] Test in a consumer that uses required props (e.g., FluxUiProps) and verify there aren't any unexpected errors
- [ ] Double-check with // debug: over_react_metrics that performance is acceptable (this lint has to process all property accesses to see if they're on props).

File: tools/analyzer_plugin/lib/src/diagnostic/unsafe_required_prop_access.dart [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    'componentWillReceivePropsWithContext',
+    'componentWillUpdate',
+    'getDerivedStateFromProps',
+    'getSnapshotBeforeUpdate',
+    'shouldComponentUpdate',
+    'shouldComponentUpdateWithContext',
+  }.contains(lifecycleMethod.name.lexeme);
+}
+
+//
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/diagnostic/unsafe_required_prop_access.dart]: Oops, missed this one. Added in 71431c72ec5154896dd646c5c21df38f919af702.

File: tools/analyzer_plugin/lib/src/diagnostic/unsafe_required_prop_access.dart [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+  return false;
+}
+
+final _memoAreEqualNamePattern = RegExp(r'[aA]reEqual');
+
+/// Returns whether [target] is an identifier referencing an argument to safe utility methods.
+///
+/// For example, in `props.containsProp((p) => p.requiredProp)`,
+/// returns `true` for the `p` that's followed by `.requiredProp`.
+bool isPropsSafeUtilityMethodArg(Expression target) {
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/diagnostic/unsafe_required_prop_access.dart]: Oh nope, must have just missed this one while cleaning up. Made private in 71431c72ec5154896dd646c5c21df38f919af702.

File: tools/analyzer_plugin/test/unit/util/is_props_from_render_test.dart [hunk 3]
```diff
@@ -0,0 +1,211 @@
+// Copyright 2021 Workiva Inc.
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/test/unit/util/is_props_from_render_test.dart]: Mm good catch. It looks like we've been inconsistent about adding them in this project; there's lots of files where they're missing. I vote we come back in a follow-up and batch add them to all the missing files

**Pull request merged**